### PR TITLE
Update CMake and use native CUDA language support

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62550
* __->__ #62445

PyTorch currently uses the old style of compiling CUDA in CMake which is just a
bunch of scripts in `FindCUDA.cmake`. Newer versions support CUDA natively as
a language just like C++ or C.

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

cc @malfet @seemethere

Differential Revision: [D31503350](https://our.internmc.facebook.com/intern/diff/D31503350)

### CODE REVIEW

**Comment by malfet** [comment]
@peterbell10 this sounds great, but moving to 3.12 is a bit challenging, as cmake-3.10 is the only version available in ubuntu-18.04

File: CMakeLists.txt [hunk 1]
```diff
@@ -489,16 +503,14 @@ if(MSVC)
   endforeach(flag_var)
 
   # Try harder
-  list(APPEND CUDA_NVCC_FLAGS "-Xcompiler" "/w" "-w")
+  string(APPEND CMAKE_CUDA_FLAGS " -Xcompiler /w -w")
```
Comment (malfet) [inline CMakeLists.txt:506]: Hmm, can you please explain why do you want to go from `list` to `string`? `string(APPEND` is inherently bad if code is compiled in folder than have spaces in it.

**Comment by malfet** [comment]
When PR is finalized, please add an explanation why you are replacing `list(ADD` with `string(APPEND`, as this change was previously done to support compilation in folders with spaces
Also, some explanation would be good why all those extra `-Wno-this-or-that` are added to the code?

File: CMakeLists.txt (hunk 1 above)
Comment (peterbell10) [reply CMakeLists.txt:506]: A list is a string separated with semicolons, but `CMAKE_CUDA_FLAGS` expects a normal string so it will pass the semicolons into the command line invocation.

File: .jenkins/pytorch/build-asan.sh [hunk 2]
```diff
@@ -16,6 +16,9 @@ clang --version
 # detect_leaks=0: Python is very leaky, so we need suppress it
 # symbolize=1: Gives us much better errors when things go wrong
 export ASAN_OPTIONS=detect_leaks=0:symbolize=1:detect_odr_violation=0
+if [ -n "$(which conda)" ]; then
+  export CMAKE_PREFIX_PATH=/opt/conda
```
Comment (peterbell10) [inline .jenkins/pytorch/build-asan.sh:20]: For some reason the non-system cmake doesn't pick up the conda installed MKL without manually adding it to the prefix path.

File: .jenkins/pytorch/win-test-helpers/build_pytorch.bat [hunk 3]
```diff
@@ -96,7 +96,7 @@ if "%USE_CUDA%"=="1" (
   :: so we are actually pretending sccache instead of nvcc itself.
   curl -kL https://github.com/peterjc123/randomtemp-rust/releases/download/v0.3/randomtemp.exe --output %TMP_DIR_WIN%\bin\randomtemp.exe
   set RANDOMTEMP_EXECUTABLE=%TMP_DIR_WIN%\bin\nvcc.exe
-  set CUDA_NVCC_EXECUTABLE=%TMP_DIR_WIN%\bin\randomtemp.exe
+  set CUDA_NVCC_EXECUTABLE=%TMP_DIR%/bin/randomtemp.exe
```
Comment (peterbell10) [inline .jenkins/pytorch/win-test-helpers/build_pytorch.bat:99]: I think this is issue might be cygwin related? It was interpreting the path as unix-style and so the backslashes were parsed as escape characters.

File: cmake/Dependencies.cmake [hunk 4]
```diff
@@ -33,6 +33,50 @@ macro(enable_ubsan)
   endif()
 endmacro()
 
+# ---[ CUDA
+if(USE_CUDA)
```
Comment (peterbell10) [inline cmake/Dependencies.cmake:37]: This code is the same, but moved to the top because `thread.cmake` uses `USE_CUDA` now.

**Review by peterbell10 — COMMENTED** [review]
@malfet PTAL. Getting CI green was harder than expected but it's finally passing now.

**Comment by malfet** [comment]
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/).

**Comment by peterbell10** [comment]
@malfet ping

File: cmake/public/utils.cmake [hunk 5]
```diff
@@ -446,10 +459,18 @@ function(torch_compile_options libname)
...
     elseif(WERROR)
-      target_compile_options(${libname} PRIVATE -Wno-strict-overflow)
+      list(APPEND private_compile_options -Wno-strict-overflow)
     endif()
   endif()
 
+  target_compile_options(${libname} PRIVATE
+      $<$<COMPILE_LANGUAGE:CXX>:${private_compile_options}>)
+  if(USE_CUDA)
+    string(REPLACE ";" "," private_compile_options "${private_compile_options}")
```
Comment (malfet) [inline cmake/public/utils.cmake:469]: Can you please add a short comment explaining that this converts list into string. 
And wouldn't `foreach` loop express the intent in a clearer way in this case?

**Review by malfet — APPROVED** [review]

File: aten/src/ATen/native/sparse/cuda/SoftMax.cu [hunk 6]
```diff
@@ -392,8 +392,6 @@ void cuda_sparse_coo_softmax(
   }
 
   cudaStream_t stream = at::cuda::getCurrentCUDAStream();
-  auto allocator = THCThrustAllocator(globalContext().lazyInitCUDA());
-  auto policy = thrust::cuda::par(allocator).on(stream);
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply aten/src/ATen/native/sparse/cuda/SoftMax.cu:396]: No, this PR propagates more C++ flags to nvcc so these unused variables fail the Werror build.

File: c10/cuda/CMakeLists.txt [hunk 7]
```diff
@@ -49,9 +49,7 @@ if(${COMPILER_SUPPORTS_HIDDEN_VISIBILITY})
 endif()
 
 # ---[ Dependency of c10_cuda
-target_link_libraries(c10_cuda PUBLIC c10)
-
-target_link_libraries(c10_cuda INTERFACE torch::cudart)
+target_link_libraries(c10_cuda PUBLIC c10 torch::cudart)
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply c10/cuda/CMakeLists.txt:52]: Because `c10_cuda` does need to linked against cudart and it's a quirk of FindCUDA that this wasn't necessary before (it was linked to cudart automatically).

File: cmake/Dependencies.cmake [hunk 8]
```diff
@@ -1610,22 +1612,19 @@ if(NOT INTERN_BUILD_MOBILE)
...
 
   if(NOT CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
     set(CMAKE_CXX_STANDARD 14)
   endif()
 
-  list(APPEND CUDA_NVCC_FLAGS ${TORCH_NVCC_FLAGS})
-  if(CMAKE_POSITION_INDEPENDENT_CODE AND NOT MSVC)
-    list(APPEND CUDA_NVCC_FLAGS "-Xcompiler" "-fPIC")
-  endif()
-
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply cmake/Dependencies.cmake:1624]: Yes, I can look in `compile_commans.json` and see that `-Xcompiler=-fPIC` is in the command line even without this.

File: cmake/public/utils.cmake (hunk 5 above)
Comment (peterbell10) [reply cmake/public/utils.cmake:470]: I've added a comment to make the intention clear. I really do want a comma-separated list though, so a foreach would just end up as more code IMO.

File: cmake/public/cuda.cmake [hunk 9]
```diff
@@ -493,3 +504,7 @@ list(APPEND CUDA_NVCC_FLAGS "--expt-relaxed-constexpr")
 
 # Set expt-extended-lambda to support lambda on device
 list(APPEND CUDA_NVCC_FLAGS "--expt-extended-lambda")
+
+foreach(FLAG ${CUDA_NVCC_FLAGS})
+  string(APPEND CMAKE_CUDA_FLAGS " ${FLAG}")
+endforeach()
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply cmake/public/cuda.cmake]: Done

File: aten/src/ATen/native/sparse/cuda/SoftMax.cu (hunk 6 above)
[the remark this answers is not in the document]
Comment (malfet) [reply aten/src/ATen/native/sparse/cuda/SoftMax.cu:396]: I'm not an expert here, but aren't `policy` is something like a thread-guard/context manager (i.e. until this class is destroyed all thrust primitives are executed on the cuda stream specified by `stream` variable instead of a default one?
@ngimel can you confirm whether its the case of changes here are ok to remove?
P.S. Can you perhaps move this change into separate PR that I can land ASAP?

File: cmake/public/utils.cmake (hunk 5 above)
Comment (malfet) [reply cmake/public/utils.cmake:470]: Sounds good, but I wonder if we need to check if private_compile_options is space-free?

File: aten/src/ATen/native/sparse/cuda/SoftMax.cu (hunk 6 above)
[the remark this answers is not in the document]
Comment (peterbell10) [reply aten/src/ATen/native/sparse/cuda/SoftMax.cu]: You can see from other parts of this file that the execution policy is passed explicitly to the function:
https://github.com/pytorch/pytorch/blob/e83d0a606ad3b72e7054ac56f04011226d6bca31/aten/src/ATen/native/sparse/cuda/SoftMax.cu#L244-L245

So, the fact it compiles without this variable proves that it's dead code. You can also see the function doesn't actually use thrust directly, instead it's all in `compute_pool_max` which defines it's own `policy`.

I've moved this change out into  #65539.

**Comment by peterbell10** [comment]
@malfet any updates?

**Review by malfet — APPROVED** [review]
